Add visual indicator if a block is connected to block binding source

michalczaplinski opened the pull request:
⚠️ This PR will need a slight refactor once https://github.com/WordPress/gutenberg/pull/58895 is merged. Once that's done, we'll have proper mechanism to add logic to "bound" blocks.

## What?
Add visual indicators if a block is connected to block bindng sources:
1. A "connection" icon in the block toolbar.
2. Focus color of the bound block is now `#9747FF`

Part of fixes in https://github.com/WordPress/gutenberg/issues/58978.

## Why?
To make the "bound" blocks visually distinct.

## How?
- Adding a new icon to `@wordpress/icons`
- Add logic to the hook which handles block binding registration.

## Testing Instructions
Register a new custom field however you prefer. You can use a snippet similar to this:

```PHP
register_meta(
	'post',
	'text_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'This is the content of the text custom field',
	)
);
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

In all of the following cases, verify that the purple focus outline is visible and that the connection icon appears both in the block toolbar and the list view.

**_Test paragraph_**
Add a paragraph with the content connected to a custom field:

```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_custom_field"}}}}} -->
<p>Content of the PAGE text custom field</p>
<!-- /wp:paragraph -->
```

**_Test heading_**
Repeat the paragraph test but using a heading.

**_Test button_**
Add a button with the content connected to a custom field:
```
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"metadata":{"bindings":{"text":{"source":"core/post-meta","args":{"key":"text_custom_field"}}}}} -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Test</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->
```

## Screenshots or screencast <!-- if applicable -->

<img width="1214" alt="Screenshot 2024-02-27 at 8 37 55 PM" src="https://github.com/WordPress/gutenberg/assets/5360536/1d3fbd44-4b5e-405b-97c3-3a545f4097e4">

michalczaplinski committed:
Add BlockControlsFirst slot to block controls groups

michalczaplinski committed:
Add connection icon to BlockControls toolbar button

michalczaplinski committed:
Add block binding toolbar button if block is connected to a source

Mamaduka commented:
The release beta cycle might not be a good time to introduce a new public SlotFill.

We can use a private SlotFill (see example #49819) or add a component directly to the toolbar like `BlockParentSelector`. What do you think?

gziolo commented:
> The release beta cycle might not be a good time to introduce a new public SlotFill.
> 
> We can use a private SlotFill (see example #49819) or add a component directly to the toolbar like `BlockParentSelector`. What do you think?

This is the expected change in the toolbar that triggered refactoring to introduce a new slot:

> Add a "bound" icon in the block toolbar.
> 
> ![Screenshot 2024-02-13 at 17 00 38](https://github.com/WordPress/gutenberg/assets/34552881/26f1cdbc-5977-4bdb-be87-2a12354c44bf)

@SaxonF or @jasmussen, what would be the alternative to change the block's icon instead in a way that retains the original but also gives a clear indicator that it's connected? This way, we wouldn't have to introduce another toolbar group.

Anyway, the other aspect is that we should stop using the filters to integrate Block Bindings with the block editor, but instead consider it part of the framework that should be tightly integrated. In effect, the changes should go directly to the block toolbar to add icon indicator there. The border for the block's content can be integrated withing `useBlockProps`, etc.

gziolo commented:
> This PR will need a slight refactor once https://github.com/WordPress/gutenberg/pull/58895 is merged. Once that's done, we'll have proper mechanism to add logic to "bound" blocks.

We will need to cherry-pick this PR to the `wp/6.5` branch, so I would wait to merge the #58895 refactoring until WP 6.5 RC1 so this bug fix lands without conflicts.

jasmussen commented:
> what would be the alternative to change the block's icon instead in a way that retains the original but also gives a clear indicator that it's connected? This way, we wouldn't have to introduce another toolbar group. In its current shape, should it even be a toolbar item? Is it actionable?

Functionally this seems okay in the name of shipping the work. That being said, it feels like there's potential for a deeper design iteration going forward, since the feature is so potent, so this is something we'll want to follow up on. The issue here goes back to the same issue we have elsewhere in the site editor, more clearly delineating what content is reusable/global, and what content is local. We haven't fully threaded the needle on how to address that; the purple sync color is a useful start, the "Content" panel and some of the work around pattern overrides can potentially be leveraged here as well. But ideally we can find one concept that would work across all of those, and make it much clearer for anyone editing, where and what they are changing.

michalczaplinski committed:
Add i18n support for block toolbar button label

michalczaplinski committed:
Add BlockBindingsButton component and remove BlockControlsFirst group

michalczaplinski commented:
Thanks for the feedback folks! I've moved the implementation directly to the `<BlockToolbar>`. My only concern with this approach is performance because then we have to run an [additional selector ](https://github.com/WordPress/gutenberg/blob/79c325d47efd1cb87c566c5207fae9a4a48fe3d9/packages/block-editor/src/components/block-bindings-button/index.js#L17-L18) for every block.

I've tried to use the private SlotFills but I wasn't quite able to make it work. Unlike the `BlockControls` SloFill, I saw duplicated elements (there was a button on the toolbar for each connected block). Perhaps they'd have to be deduplicated manually.

<img width="500" alt="Screenshot 2024-02-21 at 19 05 40" src="https://github.com/WordPress/gutenberg/assets/5417266/d2684464-0d45-4cc1-9a81-90e538c520ff">

michalczaplinski commented on packages/block-editor/src/components/block-bindings-button/index.js:
```diff
@@ -0,0 +1,36 @@
+		<ToolbarButton
+			icon={ connection }
+			label={ _x(
+				// TODO: Let's get this naming right
+				'Connected to a block bindings source',
+				'block toolbar button label'
```
Naming is hard 🤷 Happy to change it to something better.

Mamaduka replied:
Maybe just "Connected"? Toolbar button labels are usually short.

This is how the label looks when the "Show button text labels" preference is enabled.

![CleanShot 2024-02-21 at 23 17 31](https://github.com/WordPress/gutenberg/assets/240569/59f9115f-e84a-4146-906f-54657af3718c)

michalczaplinski replied:
👍 bbd3fcc5257ace5d1620e9aad2706acd6feef1ba

michalczaplinski commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -73,18 +75,28 @@ const createEditFunctionWithBindingsAttribute = () =>
+				// TODO: only set this if the block is connected to a source.
+				// this might not be a good way to do it if a block can have child
+				// blocks that are NOT connected to a source.
+				<div
+					style={
+						hasSource ? { '--wp-admin-theme-color': '#9747FF' } : {}
+					}
+				>
```
I'm not quite sure if this is the best approach to add color to the focus outline.

artemiomorales replied:
This extra containing `div` causes the in-between block inserter to break, am currently seeing how else this could be done.

michalczaplinski replied:
Right, yeah, adding that extra `div` is probably not a good idea for a few reasons. Some CSS selectors might break too.. Maybe we can add the `style` directly to `props`, sth like:

```js

props.style  = { ...props.style, '--wp-admin-theme-color': '#9747FF' };

return (
    <BlockEdit
		    key="edit"
			  { ...props }
			 attributes={ updatedAttributes }
     />
);
```

? 🤔

gziolo replied:
Having a wrapping `div` is a blocker here because it will impact existing CSS styles. Additional styles or having it abstracted as a class name can be done inside `useBlockProps`.

By the way, failing e2e tests reported on CI might be also caused by this change or by the fact that there is now an additional item in the block toolbar proposed.

artemiomorales replied:
@gziolo I've moved the styles to `useBlockProps` in https://github.com/WordPress/gutenberg/pull/59185/commits/3bc3062915c301a27cd3eb5f497df6cc3d3f1a19.

How does this look?

Mamaduka commented on packages/icons/src/library/connection.js:
```diff
@@ -0,0 +1,37 @@
+		<Path d="M5 19L8 16L5 19Z" fill="#9747FF" />
+		<Path d="M16 8L19 5L16 8Z" fill="#9747FF" />
+		<G
+			stroke-width="1.5"
+			stroke-linecap="round"
+			stroke-linejoin="round"
```
These values are producing React warnings -  `Warning: Invalid DOM property `stroke-width`. Did you mean `strokeWidth`?`

michalczaplinski replied:
Right, exactly 🙂  done in bd6e7f3ff7e2fc620d0f5c4d1ac060b459355a94

Mamaduka commented on packages/block-editor/src/components/block-bindings-button/index.js:
```diff
@@ -0,0 +1,36 @@
+import { store as blockEditorStore } from '../../store';
+
+export default function BlockBindingsButton( { clientId } ) {
+	const { blockAttributes } = useSelect(
+		( select ) => {
+			const { getBlockAttributes } = select( blockEditorStore );
+			return {
+				blockAttributes: getBlockAttributes( clientId ),
+			};
+		},
+		[ clientId ]
+	);
```
Suggestion: It's usually better to derive boolean values inside the `mapSelect`, so the editor doesn't rerender component everytime the attributes change.

Example:

```jsx
export default function BlockBindingsButton( { clientId } ) {
	const isConnected = useSelect(
		( select ) => {
			const attributes =
				select( blockEditorStore ).getBlockAttributes( clientId );

			return !! attributes?.metadata?.bindings;
		},
		[ clientId ]
	);

	return isConnected ? (
		<ToolbarButton
			icon={ connection }
			label={ _x(
				// TODO: Let's get this naming right
				'Connected to a block bindings source',
				'block toolbar button label'
			) }
			iconSize={ 24 }
		/>
	) : null;
}
```

michalczaplinski replied:
Totally 👍 585d5a3ef55272bb370d90b6dfd747bdd34dace4

Mamaduka reviewed (commented):
> My only concern with this approach is performance because then we have to run an additional selector for every block.

That should be okay in this case. The block toolbar is only rendered for the currently selected block.

SaxonF commented:
Wondering if we could highlight properties in the toolbar that are connected much like we will do in inspector where we use an icon. In toolbar we could highlight with colour + shape.  If a block's "primary" attribute is connected we'd highlight the block icon e.g. a paragraph content.

![image](https://github.com/WordPress/gutenberg/assets/1072756/29516a96-96be-42b8-973c-39828ba51b1e)

So for an image that has src, link and alt text connected when in a template it would look something like

![image](https://github.com/WordPress/gutenberg/assets/1072756/d4af98e0-eec7-411f-a453-694ba1e73abb)

michalczaplinski committed:
Refactor BlockBindingsButton to check for block connections

afercia commented:
Any important information need a visual indicaotr but also needs to be provided in a semantic and accessible way. I'm not sure a button that doesn't do anything and its only used for visual purposes is ideal.

- The new ToolbarButton doesn't do anything.
- Toolbar are supposed to contain controls for actual actions. Toolbars aren't the right place for descriptions.
- The label of the ToolbarButton is too long as @Mamaduka reported. More importantly, labels and tooltips of buttons are meant to expose the button _name_, which is what the button does. Besides the fact this button doesn;;t do anything the text `Connected to a block bindins source` or even a horter version of it is a _description_ of one of the block properties, not really the anme of the button action.
- Color alone can't be the only indication of the connected state.
- Consistency: the first button within the toolbar is supposed to:
  - Indicate the block type
  - Provide the transform to menu, which is communciated via the button description.
- Please, no purple or blue 'dots' to indicate a state: we already tried them in teh templates list and they're not great for accessibility. Also, toolbar buttons aren't the right place to indicate properties that belong to the block.

Overall, I'm not sure the toolbar is the right place to add an indicator to. I'd see the Settings panel a sa more appropriate palce, as that's where block properties and settings are available to users.

michalczaplinski commented:
Interesting concept @SaxonF, thanks! That does look better visually IMO.

My only concern is that we would have to use custom fields-specific logic to handle this. Sources other than custom fields might not have fields whose names we can display here.

<img width="209" alt="Screenshot 2024-02-22 at 13 29 34" src="https://github.com/WordPress/gutenberg/assets/5417266/73199633-d036-44f0-9ac8-09e516adb8ce">


What if it just says "Connected to `core/post-meta`" ?

michalczaplinski commented:
> Interesting concept @SaxonF, thanks! That does look better visually IMO.
> 
> My only concern is that we would have to use custom fields-specific logic to handle this. Sources other than custom fields might not have fields whose names we can display here.
> 
> <img alt="Screenshot 2024-02-22 at 13 29 34" width="209" src="https://private-user-images.githubusercontent.com/5417266/307010966-73199633-d036-44f0-9ac8-09e516adb8ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2MTAzNTIsIm5iZiI6MTcwODYxMDA1MiwicGF0aCI6Ii81NDE3MjY2LzMwNzAxMDk2Ni03MzE5OTYzMy1kMDM2LTQ0ZjAtOWFjOC0wOWU1MTZhZGI4Y2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMTM1NDEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDA2NjVmOTkwOGM0YWMxNDQ5ZmRhODczZTIzM2E2NzQyMjk5OTAwMjI2ZTViODVmNGE2NzgyOTE5ZjNkOWY3MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8mVbSbbJYpigzVy4eTkXFcwEMhWQ71tiQIQS_Hv850k">
> What if it just says "Connected to `core/post-meta`" ?

Actually, scratch that because each of block's attributes can be connected to multiple sources so that won't work either 🙁 .

What we'd need is a way to indicate the "Connected to <xyz>" in the UI _per attribute_

michalczaplinski committed:
Change the ToolbarButton label

michalczaplinski commented:
Hi @afercia 

I understand some of your concerns.

About the "button": We could get rid of the `<ToolbarButton>` and just use a `<div>` with an icon in it. What would you propose to use instead?

-----

> Toolbar are supposed to contain controls for actual actions. Toolbars aren't the right place for descriptions.

That's a subjective opinion about what the UI should look like so it's hard to argue one way or the other 🙂 

----

> The label of the ToolbarButton is too long 

I've changed it to just "Connected" in bbd3fcc5257ace5d1620e9aad2706acd6feef1ba

-----

> Color alone can't be the only indication of the connected state.

I'm not sure I understand - It's not the only indicator, it's **in addition** to the indicators that are already present:
- block editing is disabled
- there's a icon in the toolbar

So, it's an improvement over the status quo.

----

> Consistency: the first button within the toolbar is supposed to:
> - Indicate the block type
> - Provide the transform to menu, which is communciated via the button description.

That's again a subjective opinion about what UI should look like 🙂 Besides, e.g. the Button block has a parent selector as the first item in the toolbar.

----

> Please, no purple or blue 'dots' to indicate a state: we already tried them in teh templates list and they're not great for accessibility. Also, toolbar buttons aren't the right place to indicate properties that belong to the block.

As of right now, there is no indication of the state so I don't understand how adding dots would be worse? The design proposed by @SaxonF also adds indicators of "connected" state in the block Settings as you suggest. Is that what you're proposing?

michalczaplinski committed:
Update block-bindings-button import to block-bindings-indicator

michalczaplinski commented:
RE: Avoiding rendering a `<button>`. I'm not 100% sure how to proceed here. Here's what I tried:

1. Keep `<ToolbarButton>` but find a way to pass the `as` prop to render a different component like `<ToolbarButton as={"div"}>`. The `<ToolbarButton>` does not support the `as` prop.

2. Use the more generic [`ToolbarItem` component](https://developer.wordpress.org/block-editor/reference-guides/components/toolbar-item/). This is the current approach I tried in 06fdbf7ae2dfdfdae37c556d8bc8a3e400f8fa71. However, using the `ToolbarItem` also comes with a disadvantage of not having a `label` and `tooltip` as a button does.

artemiomorales commented:
This works for me — I can see the UI as expected 🙌

Regarding the design questions and accessibility ideas, that's a tough nut to crack. But if there's a concern about the toolbar buttons being actionable or not and consistency with the rest of the UI, what if we were to 1.) instead put the button towards the end of the toolbar, and 2.) make it so this button shows a popup of the currently connected attributes, and/or allows a user to remove the bindings?

<img width="622" alt="connected-block-ideas2" src="https://github.com/WordPress/gutenberg/assets/5360536/7df34b93-8c3c-469d-a167-d5ce120504d3">

I believe this was in a mockup shared earlier by @SaxonF and not sure if adding Step 2 would be a good move for 6.5, but if our intention is to eventually use this toolbar item as a button to configure bindings anyway, maybe actually making it a button and giving it the bare minimal functionality could be a good first step.

artemiomorales commented:
> I'm not sure I understand - It's not the only indicator, it's in addition to the indicators that are already present:
>- block editing is disabled
>- there's a icon in the toolbar
>
>So, it's an improvement over the status quo.

I also [added an icon in the list view](https://github.com/WordPress/gutenberg/pull/59331), which could help with providing an additional indicator but we can feel free to take or leave if it's too much for the moment.

artemiomorales commented:
> I also https://github.com/WordPress/gutenberg/pull/59331, which could help with providing an additional indicator but we can feel free to take or leave if it's too much for the moment.

As per a conversation with @SaxonF, it seems we may want to start small and avoid adding too many icons to the UI right away, so will close this for now.

artemiomorales committed:
Block Bindings: Add connection icon to list view (#59331)

* Add connection icon to list view

* Remove extraneous string

gziolo commented on packages/block-editor/src/components/block-bindings-indicator/index.js:
```diff
@@ -0,0 +1,36 @@
+import { store as blockEditorStore } from '../../store';
+
+export default function BlockBindingsIndicator( { clientId } ) {
+	const isConnected = useSelect(
```
There is now more convenient and performant way to access block bindings for the block:

https://github.com/WordPress/gutenberg/blob/7eb0a5acb3f8ab7e8edb30e26934e25477fb1b67/packages/block-editor/src/components/rich-text/index.js#L121-L123

artemiomorales replied:
@gziolo  This block bindings aren't readily present in the block edit context for the Block Bindings Indicator component, unfortunately.

I believe this is because the context is is created in [Block Edit](https://github.com/WordPress/gutenberg/blob/fix/update-block-bindings-ui-indicator/packages/block-editor/src/components/block-edit/index.js), which is instantiated in [Block List](https://github.com/WordPress/gutenberg/blob/fix/update-block-bindings-ui-indicator/packages/block-editor/src/components/block-list/index.js), whereas the Block Bindings Indicator is found outside of that scope in the block toolbar — though I may be misunderstanding here, or perhaps it's possible to pass the context to the indicator somehow.

Should I look into trying to make the context available here and/or keep digging into this?

gziolo replied:
Yes, that makes sense since this new toolbar item gets added directly to the block toolbar. All other toolbar items are usually injected as fills so they have access to the block edit context. I missed that important nuance, so it's perfectly fine as is.

SantosGuillamot replied:
I believe we could/should move the logic to check if the block is connected to [the toolbar component](https://github.com/WordPress/gutenberg/pull/59185/files#diff-9ac2a4466e7c0a85aaff0b40c5d7bc64c894132528ca3c879db4bebb438617dcR170). It is the way the rest of components seems to be used.

I mean exporting this component as it is, with the Toolbar Item, and load it conditionally adding a `isUsingBindings` or something similar to [the Toolbar useSelect logic](https://github.com/WordPress/gutenberg/blob/9d6f97c42fcc5447f0e1b38bc413ab3db9513cc5/packages/block-editor/src/components/block-toolbar/index.js#L60-L113). That way, we could just load it conditionally by doing:

```js
{ isUsingBindings && <BlockBindingsIndicator /> }
```

What do you think?

gziolo commented on packages/block-editor/src/components/list-view/block-select-button.js:
```diff
@@ -75,6 +81,8 @@ function ListViewBlockSelectButton(
 	const images = useListViewImages( { clientId, isExpanded } );
 	const { rootClientId } = useListViewContext();
 
+	const isConnected = getBlockAttributes( clientId )?.metadata?.bindings;
```
Same as above, it can be now accessed through the block edit context.

artemiomorales replied:
Similar to [this comment](https://github.com/WordPress/gutenberg/pull/59185#discussion_r1505200130), I'm not sure the context can be readily passed to the list view component, though I may be misunderstanding here.

gziolo replied:
It's me who misunderstood how the components are composed 😄

gziolo reviewed (changes requested):
See my feedback about failing e2e tests that need to be fixed before proceeding further.

artemiomorales committed:
Move bindings style to useBlockProps

artemiomorales committed:
Remove extraneous comment

gziolo commented on packages/block-editor/src/components/block-bindings-indicator/index.js:
```diff
@@ -0,0 +1,36 @@
+ */
+import { store as blockEditorStore } from '../../store';
+
+export default function BlockBindingsIndicator( { clientId } ) {
```
How about making the name more explicit? Example:
```suggestion
export default function BlockBindingsToolbarIndicator( { clientId } ) {
```

If folks agree, then the file name and class name change would have to follow.

SantosGuillamot commented on packages/icons/src/library/connection.js:
```diff
@@ -0,0 +1,37 @@
+ */
+import { SVG, Path, G } from '@wordpress/primitives';
+
+const connection = (
```
From what I can see, all the icons in the package aren't filled: [link](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library). Would it make sense to export the icon without fill, just in case someone else wants to use it, and fill it with CSS in our specific use case?

michalczaplinski replied:
Yes, I think that's the right approach 👍

artemiomorales replied:
> Would it make sense to export the icon without fill, just in case someone else wants to use it, and fill it with CSS in our specific use case?

@SantosGuillamot When you say _fill_, do you mean the purple stroke? I've [pushed a commit ](https://github.com/WordPress/gutenberg/pull/59185/commits/d571f82ef0b6797c3a2e8ab07d13e5e261f5810e) that modifies the SVG so that it's black and white by default like the other icons in the design library, and we now add the purple color using CSS.

SantosGuillamot replied:
Yes, when I said `fill` I meant they don't have the `fill` attribute to set the color.

gziolo reviewed (approved):
It's looking good code-wise. I don't have enough expertise to tell whether using `--wp-admin-theme-color` is the best approach here. I also left one [non-blocking suggestion](https://github.com/WordPress/gutenberg/pull/59185/files#r1505566084) regarding the new component name.

We should confirm the current state of visual changes with the design team.

SantosGuillamot commented:
During my testing I've seen that it is adding the visual indicators when I bind a block that is not supported. Is this the expected behavior?

SantosGuillamot commented on packages/block-editor/src/components/block-toolbar/index.js:
```diff
@@ -166,6 +167,7 @@ export function PrivateBlockToolbar( {
 				{ ! isMultiToolbar &&
 					isLargeViewport &&
 					isDefaultEditingMode && <BlockParentSelector /> }
+				<BlockBindingsIndicator clientId={ blockClientId } />
```
I believe we could move the conditional here as the rest of the component. See my other comment: [link](https://github.com/WordPress/gutenberg/pull/59185/files#r1505557629).

gziolo replied:
That's a very good idea. It would allow us to get rid of the `useSelect` call in the child component.

artemiomorales replied:
Done in https://github.com/WordPress/gutenberg/pull/59185/commits/ec3885dcca2e3c851f6b1dd3a2461a8f8e300c91

artemiomorales committed:
Move bindings selector logic to toolbar

michalczaplinski commented:
> During my testing I've seen that it is adding the visual indicators when I bind a block that is not supported. Is this the expected behavior?

That's a good catch @SantosGuillamot . In addition to checking `isConnected` we should also check if the block can be bound using [`canBindBlock()`](https://github.com/WordPress/gutenberg/blob/ab2123d24607849b49606fd1d800d8df8a3b96e8/packages/block-editor/src/hooks/use-bindings-attributes.js#L40)

artemiomorales committed:
Rename indicator file

artemiomorales committed:
Move purple stroke style from SVG markup to CSS

artemiomorales committed:
Merge branch 'trunk' into fix/update-block-bindings-ui-indicator

artemiomorales committed:
Check if block can be bound before adding styles

artemiomorales commented:
> That's a good catch @SantosGuillamot . In addition to checking isConnected we should also check if the block can be bound using [canBindBlock()](https://github.com/WordPress/gutenberg/blob/ab2123d24607849b49606fd1d800d8df8a3b96e8/packages/block-editor/src/hooks/use-bindings-attributes.js#L40)

@michalczaplinski Addressed this in https://github.com/WordPress/gutenberg/pull/59185/commits/ac72c0a108631f1f41060ef77a5c96b7333d2563

gziolo commented on packages/block-editor/src/components/block-bindings-toolbar-indicator/index.js:
```diff
@@ -0,0 +1,20 @@
+			<ToolbarItem
+				as={ 'div' }
+				aria-label={ _x( 'Connected', 'block toolbar button label' ) }
+				className="block-editor-block-bindings-indicator block-toolbar-block-bindings-indicator"
```
I checked the guidelines for [CSS naming for React components](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#naming) so it would better fit to have:

```suggestion
				className="block-editor-block-bindings-toolbar-indicator"
```

I don't know how much this pattern is being followed in practice these days. We definitely need only one class name here.

michalczaplinski replied:
I'm being pedantic here but reading the guideline, shouldn't it be `block-editor-block__bindings-toolbar-indicator` ?

> Any descendent of the component's root element must append a dash-delimited descriptor, separated from the base by two consecutive underscores __.
>
> Root element: package-directory
> Child elements: package-directory__descriptor-foo-bar

michalczaplinski replied:
In the grand scheme of things it's not super important but I've updated it in 3f9fa3bc2622ef8af610816e3c75e1a387c159c0. 

We add some of the styles to the `BlockList` component [here](https://github.com/WordPress/gutenberg/blob/3f9fa3bc2622ef8af610816e3c75e1a387c159c0/packages/block-editor/src/components/list-view/block-select-button.js#L291-L292) as well. So having two classes makes sense here: 
-  `block-editor__block-bindings-indicator` is used in both
- `block-toolbar__block-bindings-indicator` is used only in the toolbar.

I'm not 100% sure if we should put both styles inside of https://github.com/WordPress/gutenberg/blob/3f9fa3bc2622ef8af610816e3c75e1a387c159c0/packages/block-editor/src/components/block-bindings-toolbar-indicator/style.scss but I don't see a better place for shared styles.

gziolo replied:
> Root element: package-directory

Package is `@wordpress/block-editor`.
Directory is `block-bindings-toolbar-indicator`.

The `__` part comes into play when you put additional classed as child elements or when you have more components in different files than `index.js` in the same folder.

michalczaplinski replied:
Uh, right! I misunderstood something there. I'll adjust it now.

artemiomorales replied:
We might be able to get rid of the second class — checking now

artemiomorales replied:
Never mind — your recent changes look better than what I had in mind.

michalczaplinski replied:
> I'm being pedantic here but reading the guideline, shouldn't it be `block-editor-block__bindings-toolbar-indicator` ?

Yeah, I meant to write `block-editor__block-bindings-toolbar-indicator` there, adding to the confusion.

Instead, now I used `block-editor-list-view-block-select-button__bindings` and `block-editor-block-bindings-toolbar-indicator` which should be in line with the convention in GB.

gziolo commented on packages/block-editor/src/components/block-bindings-toolbar-indicator/style.scss:
```diff
@@ -0,0 +1,10 @@
+}
+
+.block-editor-block-bindings-indicator svg g {
+	stroke: #9747ff;
```
@jasmussen or @SaxonF - do you usually put colors that are used in several places under a variable name?

michalczaplinski replied:
1. Should we reuse this property ?https://github.com/WordPress/gutenberg/blob/3f9fa3bc2622ef8af610816e3c75e1a387c159c0/packages/base-styles/_default-custom-properties.scss#L7

2. Currently the color of the "bound" blocks is `#9747ff` which is close but different from synced blocks color (`#7a00df`). Should it be the same?

michalczaplinski replied:
OK, in 4dc11032bead87178814029fd0593c0a66ce6cf0, I've added a new `--wp-bound-block-color` property to [_default-custom-properties.scss](https://github.com/WordPress/gutenberg/blob/3f9fa3bc2622ef8af610816e3c75e1a387c159c0/packages/base-styles/_default-custom-properties.scss#L7) that I reused in other components.

We can still change this color in the future.

jasmussen replied:
I'd defer to Saxon, but my instinct would've been to make these the same, so it could be good to follow up and retire the variable. We _do_ reuse colors like this, there should almost never be stray hex codes in the codebase. But at the same time we're also very careful with introducing new variables. Sorry for the delayed response @michalczaplinski

michalczaplinski committed:
Simplify the SVG icon:
- get rid of 2 unnecessary `<path>` elements
- move the stroke styles to CSS
- add the `evenodd` rule

michalczaplinski committed:
Update the CSS namespacing to include the `__`

michalczaplinski commented:
I'm pretty happy with the PR now. If there aren't any more concerns I will merge later today :)

michalczaplinski committed:
Fix issues with block binding indicator color

michalczaplinski commented:
Not sure why "some checks were not successful" because looking at the individual checks they all seem to pass. I will merge now as this seems like a github reporting issue.

There's one last thing I've noticed, the block icon color in the List View is barely visible if the block is selected:

<img width="233" alt="Screenshot 2024-02-29 at 19 30 01" src="https://github.com/WordPress/gutenberg/assets/5417266/58b9bb3e-5071-4951-8739-e09dd7ba478d">

We can fix this in a quick follow-up PR.

artemiomorales commented:
> Could you check if you are able to reproduce it and, in that case, fix it please?

@SantosGuillamot Fixed in the [following commit](https://github.com/WordPress/gutenberg/pull/59477/commits/7fc13fe3f766454f947c4edf34ed7563cec56cc0) on the same PR to fix the fill color.

youknowriad commented:
I just cherry-picked this PR to the update/packages-6.5-rc1 branch to get it included in the next release
